src/org/sosy_lab/java_smt/example/NQueens.java

```diff
@@ -0,0 +1,233 @@
+     */
+    @Nullable
+    public Boolean[][] solve(int n) throws InterruptedException, SolverException {
+        BooleanFormula[][] symbols = getSymbols();
+        List<BooleanFormula> rules = getRules(symbols, context);
+        // solve N-Queens
+        try (ProverEnvironment prover = context.newProverEnvironment(ProverOptions.GENERATE_MODELS)) {
+            prover.push(bmgr.and(rules));
+            boolean isUnsolvable = prover.isUnsat();
+            if (isUnsolvable) {
+                return null;
+            }
+
+            // get model and convert it
+            Boolean[][] solution = new Boolean[n][n] ;
+            try (Model model = prover.getModel()) {
+                for (int row = 0; row < n; row++) {
+                    for (int col = 0; col < n; col++) {
+                        solution[row][col] = getValue(symbols, model, row,col);
+                    }
+                }
+                return solution;
+            }
+        }
+    }
+}
+class NQueen extends NQueensSolver {
+    private final int n;
+    public NQueen(SolverContext context, int n) {
+        super(context);
+        this.n = n;
+    }
+
+    /**
+     * prepare symbols: one symbol for each of the N*N cells.
+     */
+    @Override
+    BooleanFormula[][] getSymbols() {
+        final BooleanFormula[][] symbols = new BooleanFormula[n][n];
+        for (int row = 0; row < n; row++) {
+            for (int col = 0; col < n; col++) {
+                symbols[row][col] = bmgr.makeVariable("q_" + row + "_" + col);
+            }
+        }
+        return symbols;
+    }
+    /*
+     * getRules is the method used to add constraints that ensure that no two queens are in the same
+     * row, column, or diagonal.
+     */
+    @Override
+    public List<BooleanFormula> getRules(BooleanFormula[][] symbols, SolverContext context) {
+        List<BooleanFormula> rules = new ArrayList<>();
+        int n = symbols.length;
+
+        // At least one queen per row
+        for (BooleanFormula[] pSymbol : symbols) {
+            List<BooleanFormula> clause = new ArrayList<>(Arrays.asList(pSymbol).subList(0, n));
+            rules.add(this.bmgr.or(clause));
+        }
+
+        // At most one queen per row
+        for (BooleanFormula[] pSymbol : symbols) {
+            for (int j1 = 0; j1 < n; j1++) {
+                for (int j2 = j1 + 1; j2 < n; j2++) {
+                    rules.add(bmgr.not(bmgr.and(pSymbol[j1], pSymbol[j2])));
+                }
+            }
+        }
+        // At most one queen per column
+        for (int j = 0; j < n; j++) {
+            for (int i1 = 0; i1 < n; i1++) {
+                for (int i2 = i1 + 1; i2 < n; i2++) {
+                    rules.add(bmgr.not(bmgr.and(symbols[i1][j], symbols[i2][j])));
+                }
+            }
+        }
+        // At most one queen per diagonal
+        int numDiagonals = 2 * n - 1;
+        BooleanFormula[][] diagonals1 = new BooleanFormula[numDiagonals][n];
```

Comment:
Having an NxN square for the symbols and then two additional rectangles for just computing diagonals... this is some small overhead from a developer's point of view. However it is negligible against starting an SMT solver. Just saying. :-)

Reply:
Actually, I mostly used to do competitive programming(still took this long to solve the nqueen's problem :(  ) where comments certainly don't help due to time constraints although I am trying to improve on this. Same goes with naming variables, I am just used to naming variables with anything that comes up in my mind at that instant, which certainly is a bad habit and I would try my best to improve on it.